src/momentum.rs

```diff
@@ -1375,40 +1925,348 @@ impl Neg for Sign {
+        Signature(iter.into_iter().collect())
+    }
+}
+
+impl FromIterator<i8> for Signature {
+    fn from_iter<I: IntoIterator<Item = i8>>(iter: I) -> Self {
+        Signature(
+            iter.into_iter()
+                .map(|x| match x {
+                    0 => SignOrZero::Zero,
+                    1 => SignOrZero::Plus,
+                    -1 => SignOrZero::Minus,
+                    _ => panic!("Invalid value for Signature"),
+                })
+                .collect(),
+        )
+    }
+}
+
+impl FromIterator<isize> for Signature {
+    fn from_iter<I: IntoIterator<Item = isize>>(iter: I) -> Self {
+        Signature(
+            iter.into_iter()
+                .map(|x| match x {
+                    0 => SignOrZero::Zero,
+                    1 => SignOrZero::Plus,
+                    -1 => SignOrZero::Minus,
+                    _ => panic!("Invalid value for Signature"),
+                })
+                .collect(),
+        )
+    }
+}
+
+impl From<Vec<i8>> for Signature {
+    fn from(value: Vec<i8>) -> Self {
+        Signature::from_iter(value)
+    }
+}
+
+impl IntoIterator for Signature {
+    type Item = SignOrZero;
+    type IntoIter = std::vec::IntoIter<Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+impl<'a> IntoIterator for &'a Signature {
+    type Item = SignOrZero;
+    type IntoIter = std::iter::Copied<std::slice::Iter<'a, Self::Item>>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.iter().copied()
+    }
+}
+
+impl Index<usize> for Signature {
+    type Output = SignOrZero;
+    fn index(&self, index: usize) -> &Self::Output {
+        &self.0[index]
+    }
+}
+
+impl Signature {
+    pub fn validate_basis<T>(&self, basis: &[T]) -> bool {
+        self.len() == basis.len()
+    }
+
+    pub fn panic_validate_basis<T>(&self, basis: &[T]) {
+        if !self.validate_basis(basis) {
+            panic!(
+                "Invalid basis for Signature, expected length {}, got length {}",
+                self.len(),
+                basis.len()
+            );
+        }
+    }
+
+    pub fn to_momtrop_format(&self) -> Vec<isize> {
+        self.0
+            .iter()
+            .map(|x| match x {
+                SignOrZero::Zero => 0,
+                SignOrZero::Plus => 1,
+                SignOrZero::Minus => -1,
+            })
+            .collect()
+    }
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    pub fn iter(&self) -> std::slice::Iter<SignOrZero> {
+        self.0.iter()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+    pub fn apply<T>(&self, basis: &[T]) -> T
+    where
+        T: RefZero + Clone + Neg<Output = T> + AddAssign<T>,
+    {
+        // self.panic_validate_basis(basis);
+        let mut result = basis[0].ref_zero();
+        for (&sign, t) in self.0.iter().zip(basis.iter().cloned()) {
```

Comment:
I will have at least one comment for my review :)
It would be better here to do a match on `SignOrZero` and skip in the case of `Zero` to save one multiplication

Reply:
well it isn't a multiplication per se, since multiplying by a signorzero either instantiates a new zero, or negates the value or does nothing.

Reply:
But for sure I can optimise adding zero.

Reply:
Ah ok, you were even smarter than I thought then, all good!